### PR TITLE
Reimplementation of Tray user logging 

### LINKED ISSUE #54: No Playable Media/Thumbnail on Published Versions
Breaking this issue out from a larger one
----------
So I went to the traypublisher tool, grabbed a 10 second .mp4 of Big Buck Bunny and went to publish it as a Render type.
The publish went through perfectly and I saw it run ffmpeg to extract a reviewable from the .mp4, however, in ShotGrid I see no uploaded media nor path under the "Path To [...]" fields; also no Ayon ID or Sync Status?
image

_Originally posted by @matteoveglia in https://github.com/ynput/ayon-shotgrid/issues/22#issuecomment-1797927166_

### PR DESCRIPTION
**Pull request by jakubjezek001**
## Changelog Description
We are reintroducing the functionality for users to access the ShotGrid addon via the tray widget. Previously, we had transitioned to an environment variable approach; however, following careful consideration of user input, we have opted to introduce two additional login methods. Users now have the option to log in using their username and password, mirroring the original process. Additionally, a novel method has been implemented where users can log in solely with their username, subsequently connecting to the ShotGrid Script API through studio settings. This approach involves a dedicated script with tailored permissions, catering specifically to fundamental user publishing requirements.

## Additional info
This PR is also solving several bugs caused by transition from OpenPype to ayon-core. Versoins are integrated again.

## Testing notes
Yeah this is long one! There are three ways to test this PR and we need to test all of them.

### Preparation for testing
All I will be describing in Dev mode ON
1. Since server settings were upgraded you need to update addon at your `ayon-docker/addons` folder. In repository root create and copy addon package to your local `ayon-docker` (this is still old structure so you should not clone this repo to addons folder directly). Or do it with the ZIP file uploading if you want to waist your time 😄 , or docker with server is not running from local machine. 
```
python .\create_package.py --skip-zip --output [PATH TO YOUR REPO]\ayon-docker\addon
```
2. Restart your server and then configure your bundle to lead to your client folder. 
image
3. go to addon settings `ayon+settings://shotgrid/shotgrid_server` and add your server. (you can also set env `HTTPS_PROXY`)
4. in **core** addon go to `ayon+settings://core/publish/ExtractReview/profiles/0/outputs`, find h264 preset and add **shotgridreview** to tags - this will upload publishing version to your ShotGrid server.
5. Then follow other steps which are related to particular user login methods

### Configuring Local storage
You will need to have configured your ShotGrid server > Site settings > File Management before the publishing.
Since default settings are enabled and set to `primary` you need to make sure this is also correctly configured on your server.
Read more about it [here ](https://help.autodesk.com/view/SGSUB/ENU/?guid=SG_Administrator_ar_data_management_ar_linking_local_files_html)

### Publishing test steps
1. Make sure your Local storage settings are in sync with ShotGrid File Management presets. Look above for more info. **Configuring Local storage**.
2. Then open any host or Traypublisher and publish some sequence of video with **review** activated
3. All plugins should work well and version should be published to SG as expected

- [x] default Env method
    1. in `ayon+settings://shotgrid/client_login/env/client_sg_script_key` write the api key of ShotGrid script 
    2. in `ayon+settings://shotgrid/client_login/env/client_sg_script_name` write the name of ShotGrid script name 
    3. Save settings and you are good to open launcher in dev mode
    4. Make sure that `AYON_SG_USERNAME` environment variable is set to your SG registered email.
    5. Just in case the environment variable is correctly registered under your console either echo it or reboot the terminal.
    6. Follow **Publishing test steps**
    
- [x] User logs with password via Tray
    1. In settings `ayon+settings://shotgrid/client_login/type` set it to **Via Tray App with password**
    2. Save settings and Reboot your ayon launcher.
    3. input dialogue window should popup asking you for user@name and password you are using on SG
    4. Confirm and the window should automatically close after it is successfully registered
    5. Follow **Publishing test steps**

- [x] User logs only with name via Tray. 
    1. in settings go to `ayon+settings://shotgrid/client_login/type` and set it to **Via Tray App with shared api key**
    2. Go to your ShotGrid server and make new Script api access point for User restricted permissions. 
    3. Copy generated key and paste it to settings `ayon+settings://shotgrid/client_login/tray_api_key/client_sg_script_key` 
    4. in settings `ayon+settings://shotgrid/client_login/tray_api_key/client_sg_script_name` add the name of SG script. 
    5. Save the settings and reboot the launcher. 
    6. Dialogue window should popup with only _user@name_ to be added. (If it is not happening, go to Tray ShotGrid menu and click on change)
    7. Confirm the change and if all is successfully registered then the window should close.
    8. Follow **Publishing test steps**
 

closes #54

### CODE REVIEW

**Comment by fabiaserra** [comment]
> python .\create_package.py --skip-zip --output [PATH TO YOUR REPO]\ayon-docker\addon

What's the point of doing this? how is that used in the dev mode? I have been deving locally for a long time without needing to do that? For service changes I just rebuild the dockers using the Makefile `build` rule and run it with the `dev` rule and if there's any server changes I upload the .zip.

> in ayon+settings://shotgrid/server_sg_script_name write the name of ShotGrid script name

If you are going to use secrets, why not use the name of the secret directly as it was? Although aren't you just going to remove using secrets?

> You will need to have configured your ShotGrid server > Site settings > File Management before the publishing.
Since default settings are enabled and set to primary you need to make sure this is also correctly configured on your server.
Read more about it [here](https://help.autodesk.com/view/SGSUB/ENU/?guid=SG_Administrator_ar_data_management_ar_linking_local_files_html)

AFAICT ShotGrid automatically picks up the local storage root if the path of the file published is one of the roots in your settings so you don't need to explicitly set the local storage when you publish, it simply needs to be a path with one of those, which is why I made the local storage setting optional here https://github.com/ynput/ayon-shotgrid/pull/68

> in core addon go to ayon+settings://core/publish/ExtractReview/profiles/0/outputs, find h264 preset and add shotgridreview to tags - this will upload publishing version to your ShotGrid server.

In my fork of ayon-core I have multiple places where I hard-code the `shotgridreview` tag on the representation creations that I want to publish to ShotGrid (i.e. collect_writes.py of Nuke, my custom Houdini collectors and batch publishes...) as what you describe would only publish the generated review and not the .exrs as path to frames for example. You guys should expose the custom tags of all the collectors so users can do that.

> Make sure that AYON_SG_USERNAME environment variable is set to your SG registred email.

This PR https://github.com/ynput/ayon-core/pull/379 is also important to make this functionality work in Deadline

**Comment by jakubjezek001** [comment]
> > python .\create_package.py --skip-zip --output [PATH TO YOUR REPO]\ayon-docker\addon
> 
> What's the point of doing this? how is that used in the dev mode? I have been deving locally for a long time without needing to do that? For service changes I just rebuild the dockers using the Makefile `build` rule and run it with the `dev` rule and if there's any server changes I upload the .zip.

This is optional, you can of course do it with zip upload.. this saves a lot of time.. Try it.

**Comment by jakubjezek001** [comment]
> > in ayon+settings://shotgrid/server_sg_script_name write the name of ShotGrid script name
> 
> If you are going to use secrets, why not use the name of the secret directly as it was? Although aren't you just going to remove using secrets?

This had been already refactored and I will update description.

**Comment by jakubjezek001** [comment]
> > You will need to have configured your ShotGrid server > Site settings > File Management before the publishing.
> > Since default settings are enabled and set to primary you need to make sure this is also correctly configured on your server.
> > Read more about it [here](https://help.autodesk.com/view/SGSUB/ENU/?guid=SG_Administrator_ar_data_management_ar_linking_local_files_html)
> 
> AFAICT ShotGrid automatically picks up the local storage root if the path of the file published is one of the roots in your settings so you don't need to explicitly set the local storage when you publish, it simply needs to be a path with one of those, which is why I made the local storage setting optional here #68

I had included this step because it was not working on my side until I had configured it that way.

**Comment by jakubjezek001** [comment]
> > > > python .\create_package.py --skip-zip --output [PATH TO YOUR REPO]\ayon-docker\addon
> > > 
> > > 
> > > What's the point of doing this? how is that used in the dev mode? I have been deving locally for a long time without needing to do that? For service changes I just rebuild the dockers using the Makefile `build` rule and run it with the `dev` rule and if there's any server changes I upload the .zip.
> > 
> > 
> > This is optional, you can of course do it with zip upload.. this saves a lot of time.. Try it.
> 
> by simply having an `addon` folder and making the bundle point to the `./client` folder it finds the `addon` contents for the server package?

This only points to client code but not working in case you want to reflect the server setting changes.

**Comment by fabiaserra** [comment]
> I had updated description so it makes more sense, why is this step needed. Sorry for the confusion.

I see, I think the zip upload is actually easier to understand if you are not running a local ayon docker and it actually is easier to follow and really fast! But ideally the same `create_package` would have an `--upload` flag taking the local `.env` so it uploads the zip directly to the server like there is for the launcher

File: client/ayon_shotgrid/addon.py [hunk 1]
```diff
@@ -1,36 +1,81 @@
 import os
 
-import ayon_api
-
-from openpype.modules import (
+from ayon_core.modules import (
```
Comment (iLLiCiTiT) [inline client/ayon_shotgrid/addon.py:3]: Why is this PR combining ayon_core conversion and login via dialog?

**Comment by iLLiCiTiT** [comment]
Why is this PR combining ayon_core conversion and login via dialog?

**Comment by fabiaserra** [comment]
> Why is this PR combining ayon_core conversion and login via dialog? Those are 2 separate tasks.... And I see potential different changes unrelated to this PR. Can we split it? It is really hard to do review of 2 (and maybe more - using secrets changed too) huge changes in codebase.

I think you guys are way too picky on the granularity of PRs, disencouraging contributors to create bigger refactoring of codebases and complicating development. FWIW I did split those changes in individual PRs (https://github.com/ynput/ayon-shotgrid/pull/80) because I know it's the first thing you guys always complain about but this repo still needs a lot of love to get it into production so bigger PRs should be expected because a lot of times splitting each change in a branch slows down testing and iterations way more than the benefits of working on a single refactor. I really don't believe this PR is that big to require separating it. On the contrary, as I already warned on this PR before the merge https://github.com/ynput/ayon-shotgrid/pull/77 (and no one cared) that should have never been merged into any existing PRs because style changes DO distract a lot from reviewing proper changes and create a lot of merge conflicts

File: client/ayon_shotgrid/plugins/publish/validate_shotgrid_user.py [hunk 2]
```diff
@@ -20,49 +20,28 @@ def process(self, context):
             raise PublishValidationError("Missing Shotgrid Credentials")
 
         self.log.info("Login Shotgrid set in Ayon is {}".format(user_login))
-        self.log.info("Current Shotgrid Project is {}".format(sg_project))
+        self.log.info("Current shotgun Project is {}".format(sg_project))
```
[the remark this answers is not in the document]
Comment (jakubjezek001) [reply client/ayon_shotgrid/plugins/publish/validate_shotgrid_user.py:24]: haha! yeah it was just nostalgia 😄

**Comment by jakubjezek001** [comment]
> Are you planning on removing the use of the secret for this PR?

it is already removed for client login. Services can still use secret resolving

File: server/settings/main.py [hunk 3]
```diff
@@ -60,20 +60,43 @@ def get_default_folder_attributes():
...
         title="How often (in seconds) to process ShotGrid related events.",
         validate_default=False,
     )
 
+    script_key: str = SettingsField(
+        default="",
+        enum_resolver=secrets_enum,
+        title="ShotGrid's Script api key",
+        description=(
+            "Ayon Secret used for Service related server operations "
```
Comment (iLLiCiTiT) [inline server/settings/main.py:81]:
```suggestion
            "AYON Secret used for Service related server operations "
```

File: server/settings/main.py [hunk 4]
```diff
@@ -60,20 +60,43 @@ def get_default_folder_attributes():
...
+            "authentication.html#setting-up-shotgrid"
+        ),
+    )
+
+    script_name: str = SettingsField(
+        default="",
+        placeholder="Create and Paste a script name here",
+        title="ShotGrid's Script Name",
+        description=(
+            "Ayon Secret used for Service related server operations "
```
Comment (iLLiCiTiT) [inline server/settings/main.py:93]:
```suggestion
            "AYON Secret used for Service related server operations "
```

File: server/settings/main.py [hunk 5]
```diff
@@ -114,6 +137,64 @@ class ShotgridCompatibilitySettings(BaseSettingsModel):
...
 
+class ClientLoginDetailsModel(BaseSettingsModel):
+    _layout = "expanded"
+
+    client_sg_script_key: str = SettingsField(
+        default="",
+        placeholder="Create and Paste a script api key here",
+        title="Client related ShotGrid's Script api key",
+        description=(
+            "Ayon Secret used for Client related user operations "
```
Comment (iLLiCiTiT) [inline server/settings/main.py:148]:
```suggestion
            "AYON Secret used for Client related user operations "
```

File: server/settings/main.py [hunk 6]
```diff
@@ -114,6 +137,64 @@ class ShotgridCompatibilitySettings(BaseSettingsModel):
...
+            "See more at: https://developer.shotgridsoftware.com/python-api/"
+            "authentication.html#setting-up-shotgrid"
+        ),
+    )
+    client_sg_script_name: str = SettingsField(
+        default="",
+        placeholder="Create and Paste a script name here",
+        title="Client related ShotGrid's Script Name",
+        description=(
+            "Ayon Secret used for Client related user operations "
```
Comment (iLLiCiTiT) [inline server/settings/main.py:159]:
```suggestion
            "AYON Secret used for Client related user operations "
```

**Comment by fabiaserra** [comment]
> > Are you planning on removing the use of the secret for this PR?
> 
> it is already removed for client login. Services can still use secret resolving

Is it still a secret? I can't see from the code that the settings are a secret but I see you are querying it as a secret on the services?

**Review by iLLiCiTiT — APPROVED** [review]

**Pull request merged**